src/Oro/Bundle/EntityBundle/EventListener/NavigationListener.php

```diff
@@ -44,66 +55,110 @@ public function __construct(
-            /** @var ConfigProvider $entityConfigProvider */
-            $entityConfigProvider = $this->configManager->getProvider('entity');
-
-            /** @var ConfigProvider $entityExtendProvider */
-            $entityExtendProvider = $this->configManager->getProvider('extend');
-
-            $extendConfigs = $entityExtendProvider->getConfigs();
-
-            foreach ($extendConfigs as $extendConfig) {
-                if ($this->checkAvailability($extendConfig)) {
-                    $config = $entityConfigProvider->getConfig($extendConfig->getId()->getClassname());
-                    if (!class_exists($config->getId()->getClassName()) ||
-                        !$this->securityFacade->hasLoggedUser() ||
-                        !$this->securityFacade->isGranted('VIEW', 'entity:' . $config->getId()->getClassName())
-                    ) {
-                        continue;
-                    }
-
-                    $children[$config->get('label')] = array(
-                        'label'   => $this->translator->trans($config->get('label')),
-                        'options' => array(
-                            'route'           => 'oro_entity_index',
-                            'routeParameters' => array(
-                                'entityName' => str_replace('\\', '_', $config->getId()->getClassName())
-                            ),
-                            'extras'          => array(
-                                'safe_label' => true,
-                                'routes'     => array('oro_entity_*')
-                            ),
-                        )
-                    );
-                }
+        $entitiesMenuItem = $event->getMenu()->getChild('system_tab')->getChild('entities_list');
+        if (!$entitiesMenuItem) {
+            return;
+        }
+
+        if ($this->cache->contains(static::CACHE_KEY)) {
+            $children = $this->cache->fetch(static::CACHE_KEY);
+        } else {
+            $children = $this->getCustomEntitiesMenuConfig();
+            $this->cache->save(static::CACHE_KEY, $children, static::CACHE_LIFETIME);
+        }
+
+        foreach ($children as $child) {
+            $entitiesMenuItem->addChild($child['label'], $child['options']);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    protected function getCustomEntitiesMenuConfig()
+    {
+        $children = [];
+
+        /** @var ConfigProvider $entityConfigProvider */
+        /** @var ConfigProvider $entityExtendProvider */
+        /** @var ConfigProvider $groupConfigProvider */
+        $entityConfigProvider = $this->configManager->getProvider('entity');
+        $entityExtendProvider = $this->configManager->getProvider('extend');
+        $groupConfigProvider  = $this->configManager->getProvider('grouping');
+
+        $withHidden    = true; // TODO: move it to configuration
+        $extendConfigs = $entityExtendProvider->getConfigs(null, $withHidden);
```

Comment:
you do not need provider here, you can get Configs from configManager instead , e.g.
$this->configManager->getConfigs('extend', null, true)